woocommerce/woocommerce
# Improve webpack cache-busting version parameter by using file contents hash

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Closes https://github.com/woocommerce/woocommerce/issues/33337

This PR improves the cache-busting version parameter by using `?ver=<file hash>` instead of a version number, which makes it so that the client does not have to re-fetch resources that haven't changed recently.

There are two kinds of resources, one is scripts that are enqueued in PHP, and the other is chunks that are generated by webpack and loaded in the frontend.

For the former, the version parameter is added to the URL by the `wp_enqueue_script` function, we uses the `version` from the asset manifest file to get the file hash. It's possible to get file hash dynamically but it may affect performance, so I chose to use the asset manifest file. The approach is also used to load wp-includes scripts in WP core. 

For the latter, we simply change the `output` config in `webpack.config.js` to load the file with the hash.

The asset manifest file is generated by the [dependency-extraction-webpack-plugin](https://github.com/WordPress/gutenberg/blob/a04a8e94e8b93ba60441c6534e21f4c3c26ff1bc/packages/dependency-extraction-webpack-plugin/README.md#behavior-with-scripts) and contains the unique version hash calculated based on the file content. However, it only supports JS files. To support CSS files, I added a new plugin called `style-asset-plugin` to extract the CSS file hash and write it to the asset manifest file. It's just a simple modification of the `dependency-extraction-webpack-plugin`.

Please note that because the asset file is not regenerated in Webpack watch mode, you need to run `pnpm  build` to get the new hash and when `SCRIPT_DEBUG` is enabled, the version parameter is replaced with file modified time.





### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a fresh site
2. Set `define( 'SCRIPT_DEBUG', false );` in your wp-config.php file.
3. Open dev tool > network tab
4. Go to WooCommerce
5. Skip core profiler and go to WooCommerce > Home
6. Verify that the `?ver=<hash>` is set on all WooCommerce Admin JS and CSS requests, please ignore assets loaded from `/assets/` path since we're not able to use webpack to get file hash.
7. Modify the `plugins/woocommerce-admin/client/app.js` (add a console.log for example)
8. Run `pnpm run --filter @woocommerce/plugin-woocommerce build`
9. Reload the page and confirm the `<hash>` value is changed
10. Go to `/wp-admin/admin.php?page=wc-settings&tab=checkout`
11. Confirm `payment-method-promotions.js` script is loaded properly like `payment-method-promotions.js?ver=5a20ddfd6fe486acf2ba`.
12. Set `define( 'SCRIPT_DEBUG', true );` in your wp-config.php file.
13. Reload the page and confirm`?ver=<timestamp>` is set on all WooCommerce Admin JS and CSS requests (Ignore legacy assets from `/assets/` path and `/chunks`)

![Screenshot 2024-02-21 at 17 19 48](https://github.com/woocommerce/woocommerce/assets/4344253/13ba7d6c-b64d-42da-99f5-e774b83e1927)

When `SCRIPT_DEBUG` is enabled:

![Screenshot 2024-02-21 at 18 29 12](https://github.com/woocommerce/woocommerce/assets/4344253/989ea768-8f1f-4259-aab7-d53d0d4987b7)




### Changelog entry




<details>

#### Significance




#### Type




#### Message 

Improve webpack cache-busting version parameter by using file contents hash

#### Comment 

</details>
